File: app/src/test/java/com/werocksta/dagger2demo/presenter/RepoPresenterTest.java
package com.werocksta.dagger2demo.presenter;

import com.werocksta.dagger2demo.api.GithubAPI;
import com.werocksta.dagger2demo.model.RepoCollection;
import com.werocksta.dagger2demo.util.RxSchedulersOverrideRule;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;

import static org.junit.Assert.assertNotNull;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class RepoPresenterTest {

    @Mock public GithubAPI api;

    @Mock public RepoPresenter.View view;

    @Rule
    public RxSchedulersOverrideRule mRxSchedulersRule = new RxSchedulersOverrideRule();

    RepoPresenter presenter;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        presenter = new RepoPresenter(api);
        presenter.injectView(view);
    }

    @Test
    public void presenterShouldNotNull() throws Exception {
        assertNotNull(presenter);
    }

    @Test
    public void getRepoShouldDisplayListRepo() throws Exception {
        String user = "WeRockStar";
        List<RepoCollection> collections = new ArrayList<>();

        when(api.getRepo(user)).thenReturn(Observable.just(collections));
        presenter.getRepo(user);

        verify(view).loading();
        verify(view).displayRepo(collections);
        verify(view).loadComplete();
    }

    @Test
    public void getRepoErrorShouldReturnEmptyArray() throws Exception {
        String emptyUser = "";

        when(api.getRepo(emptyUser)).thenReturn(Observable.error(new Throwable()));
        presenter.getRepo(emptyUser);

        verify(view).loading();
        verify(view).displayRepo(new ArrayList<>());
        verify(view).loadComplete();
    }

    @After
    public void tearDown() throws Exception {
        presenter.onStop();
    }
}

File: app/src/test/java/com/werockstar/dagger2demo/presenter/RepoPresenterTest.kt
package com.werockstar.dagger2demo.presenter


import com.werockstar.dagger2demo.api.GithubAPI
import com.werockstar.dagger2demo.model.Repo
import com.werockstar.dagger2demo.util.RxScheduler
import io.reactivex.Observable
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import java.util.*

class RepoPresenterTest {

    @Mock private lateinit var api: GithubAPI
    @Mock private lateinit var view: RepoPresenter.View

    private lateinit var presenter: RepoPresenter

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        presenter = RepoPresenter(api, RxScheduler.rxScheduler)
        presenter.injectView(view)
    }

    @Test
    @Throws(Exception::class)
    fun `get repo should display list repository`() {
        val user = "WeRockStar"
        val collections = ArrayList<Repo>()

        `when`(api.getRepo(user)).thenReturn(Observable.just(collections))
        presenter.getRepo(user)

        verify(view).loading()
        verify(view).displayRepo(collections)
        verify(view).loadComplete()
    }

    @Test
    @Throws(Exception::class)
    fun `get repo error should return empty array`() {
        val emptyUser = ""

        `when`(api.getRepo(emptyUser)).thenReturn(Observable.error(Throwable()))
        presenter.getRepo(emptyUser)

        verify(view).loading()
        verify(view).displayRepo(ArrayList<Repo>())
        verify(view).loadComplete()
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter.onStop()
    }
}